File: wishlist.py
"""
wishlist.py — what the deck most wants to draw or fetch, given the board.

Selection effects (Brainstorm, Ponder, Frantic Search) and tutors (Gamble, ...) use
this to KEEP/FETCH the best cards instead of blind top-of-library, so the sim and the
planner's rollouts reflect real selective card advantage rather than random draws.

Priority is tiered, exactly as the pilot asked:
    Tier 1  state-specific wishes — fill the biggest gap in the engine right now
            (a payoff if none is accessible, a 2nd Krark body, a doubler, Krark's
            Thumb, a draw/mana engine the board lacks).
    Tier 2  generic engine cards — bodies, doublers, payoffs, gas.
    Tier 3  everything else (lands, interaction) ~ 0.

`card_value(state, name)` scores a single card; rank any drawn/searchable set by it.
"""
from __future__ import annotations

import cards

_PAYOFFS = frozenset({"Thassa's Oracle", "Grapeshot", "Brain Freeze"})
_BODIES = frozenset({"Krark, the Thumbless", "Sakashima of a Thousand Faces",
                     "Glasspool Mimic", "Phantasmal Image"})
_DOUBLERS = frozenset({"Veyran, Voice of Duality", "Harmonic Prodigy"})
_DRAW_ENGINES = frozenset({"Archmage Emeritus"})
_MANA_ENGINES = frozenset({"Storm-Kiln Artist", "Birgi, God of Storytelling"})
# Acceleration the pilot wants to TUTOR for by default — pieces that advance the game plan
# toward "power out the whole deck", after which the finish is trivially found. (Jeska's Will
# is the flagship ramp; scored specially below.)
_FAST_MANA = frozenset({"Sol Ring", "Mox Diamond", "Chrome Mox", "Lotus Petal",
                        "Arcane Signet", "Springleaf Drum", "Lion's Eye Diamond",
                        "Mana Vault", "Mox Amber", "Relic of Legends"})   # added 2026-06-21
_RITUALS = frozenset({"Pyretic Ritual", "Desperate Ritual", "Strike It Rich",
                      "Rite of Flame"})
# Clean finishes worth tutoring for unconditionally. Brain Freeze is NOT here: it only
# wins off a big self-mill, so it's only worth grabbing once a mill/draw engine is set up.
_CLEAR_PAYOFFS = frozenset({"Thassa's Oracle", "Grapeshot"})

# The Dualcaster combo: Dualcaster Mage + (Twinflame | Heat Shimmer) is a 2-card DETERMINISTIC
# kill (flash Dualcaster in response to the shimmer -> infinite hasty tokens). If you already hold
# one half, the other half is the single best card a tutor can find — it wins next turn — so it
# outranks generic acceleration. Spellseeker can fetch Twinflame (mv2); Imperial Recruiter fetches
# Dualcaster Mage; Gamble fetches either.
_COMBO_DUALCASTER = "Dualcaster Mage"
_COMBO_SHIMMERS = frozenset({"Twinflame", "Heat Shimmer"})


def _have(state, name: str) -> bool:
    return name in state.hand or state.has_permanent(name)


def _brain_freeze_ready(state) -> bool:
    """A clear Brain Freeze plan exists: 2+ bodies plus an engine to fuel repeated casts."""
    return state.krark_bodies >= 2 and (
        state.has_permanent("Archmage Emeritus")
        or state.has_permanent("Storm-Kiln Artist")
        or state.has_permanent("Birgi, God of Storytelling"))


def _combo_ready(state) -> bool:
    """The engine is assembled enough to power out the deck and close THIS turn-ish: 2+
    Krark bodies plus a sustaining mana/draw engine (or Breach to recur the finish). Only
    then is it worth burning a tutor on the end-game payoff — before that, Thassa's Oracle
    is dead weight you'd rather find naturally while digging."""
    return state.krark_bodies >= 2 and (
        any(state.has_permanent(m) for m in _MANA_ENGINES)
        or any(state.has_permanent(d) for d in _DRAW_ENGINES)
        or state.has_permanent("Urabrask")
        or state.has_permanent("Tavern Scoundrel")
        or state.has_permanent("Underworld Breach"))


def _ready_to_finish(state, name: str) -> bool:
    """Is it time to TUTOR `name` as the finish? Brain Freeze needs its mill plan; the
    Oracle/Grapeshot need the combo assembled."""
    if name == "Brain Freeze":
        return _brain_freeze_ready(state)
    return _combo_ready(state)

# Generic engine value (Tier 2) — the cards that are good in most spots.
_ENGINE = (_BODIES | _DOUBLERS | _DRAW_ENGINES | _MANA_ENGINES | _PAYOFFS | frozenset({
    "Krark's Thumb", "Baral, Chief of Compliance", "Underworld Breach",
    "Lion's Eye Diamond", "Jeska's Will", "Brainstorm", "Ponder", "Frantic Search",
    "Gamble", "Pyretic Ritual", "Desperate Ritual", "Strike It Rich",
    "Gitaxian Probe", "Peek", "Borne Upon a Wind", "Rite of Flame",   # added 2026-06-21 (cantrips + ritual)
    "Opt", "Consider", "Serum Visions", "Preordain",   # added 2026-06-21 (cut4 dig swap)
}))


def payoff_accessible(state) -> bool:
    """A finish is in hand, on the battlefield, or in the yard (Breach can escape it)."""
    return any(p in state.hand or state.has_permanent(p) or p in state.graveyard
               for p in _PAYOFFS)


def card_value(state, name: str, for_tutor: bool = False) -> float:
    """Higher = more wanted NOW. `for_tutor` distinguishes FETCHING (advance the plan —
    get acceleration; the end-game finish is found naturally) from KEEPING/selection
    (don't bottom your wincon). Tier-1 wishes dominate Tier-2 engine value, which
    dominates everything else."""
    score = 0.0
    is_body = name in _BODIES
    accessible = payoff_accessible(state)

    # Combo completion (top priority): holding one half of the Dualcaster combo, the missing half
    # is a guaranteed kill — fetch/keep it above everything else.
    have_dc = _have(state, _COMBO_DUALCASTER)
    have_shimmer = any(_have(state, sh) for sh in _COMBO_SHIMMERS)
    if name == _COMBO_DUALCASTER and have_shimmer and not have_dc:
        score += 120.0
    if name in _COMBO_SHIMMERS and have_dc and not have_shimmer:
        score += 120.0

    # Tier 1 — ACCELERATION: the pieces that advance the game plan, which the pilot wants
    # tutors to grab by default. A 2nd Krark body only matters once you already have one
    # (a library "body" is a clone — dead without a Krark to copy).
    if is_body and 1 <= state.krark_bodies < 2:
        score += 80.0
    if name in _DOUBLERS and state.trigger_doublers == 0:
        score += 60.0
    if name == "Krark's Thumb" and not state.has_permanent("Krark's Thumb"):
        score += 50.0
    if name in _DRAW_ENGINES and not any(state.has_permanent(d) for d in _DRAW_ENGINES):
        score += 50.0
    if name in _MANA_ENGINES and not any(state.has_permanent(m) for m in _MANA_ENGINES):
        score += 40.0
    # Ramp + fast mana: Jeska's Will is the flagship accelerant (explosive with a body out);
    # fast rocks/rituals power the turn out. These are the preferred default tutor targets.
    if name == "Jeska's Will":
        score += 60.0 if state.krark_bodies >= 1 else 35.0
    elif name in _FAST_MANA:
        score += 35.0
    elif name in _RITUALS:
        score += 20.0

    # Payoffs. Thassa's Oracle is an END-GAME finish — with a mana engine + a fully-seen
    # deck it's trivially found, so DON'T burn an early tutor on it. KEEP it when drawn
    # (never bottom the wincon), but only TUTOR a finish once the combo is assembled.
    if name in _PAYOFFS:
        if not for_tutor:                              # keeping/selection: hold the wincon
            score += 45.0 if name == "Thassa's Oracle" else 15.0
        if not accessible and _ready_to_finish(state, name):
            score += 100.0                             # engine assembled -> the finish is the best
                                                       # fetch (above a 3rd body / more acceleration),
                                                       # just under combo-completion (+120)

    # Tier 2 — generic engine value (payoffs already scored above).
    if name in _ENGINE and name not in _PAYOFFS:
        score += 10.0
        if is_body:
            score += 5.0

    # Tier 3 — lands / interaction / filler stay ~0.
    return score


def best(state, pool, k: int = 1, for_tutor: bool = False):
    """The k highest-value cards from `pool` (a list of names), best first."""
    return sorted(pool, key=lambda c: card_value(state, c, for_tutor), reverse=True)[:k]


def tutor(state, predicate):
    """Move the highest-value library card matching `predicate` into hand (the library
    is conceptually shuffled afterward). Returns the fetched card name, or None. Uses the
    tutor ranking — fetch acceleration, not the end-game finish."""
    matches = [c for c in state.library if predicate(c)]
    if not matches:
        return None
    fetched = best(state, matches, 1, for_tutor=True)[0]
    state.library.remove(fetched)
    state.hand.append(fetched)
    return fetched


if __name__ == "__main__":
    cards.load()
    from game_state import GameState, Permanent, krark_body

    # 1 body, no engine: a TUTOR should grab acceleration (2nd body / doubler / Jeska),
    # NOT the end-game Oracle (found naturally once the deck is powered out).
    s = GameState(battlefield=[krark_body("Krark, the Thumbless")])
    pool = {"Thassa's Oracle", "Sakashima of a Thousand Faces", "Veyran, Voice of Duality",
            "Jeska's Will", "Brainstorm", "Island", "Force of Will"}
    tut = sorted(pool, key=lambda c: card_value(s, c, for_tutor=True), reverse=True)
    print("1 body, no payoff -> TUTOR ranking:", tut)
    assert tut[0] != "Thassa's Oracle", "tutor should not grab the end-game Oracle first"
    assert card_value(s, "Thassa's Oracle", for_tutor=True) < card_value(s, "Jeska's Will", for_tutor=True)
    assert card_value(s, "Island") == 0.0
    # Keeping/selection still values the Oracle (don't bottom the wincon).
    assert card_value(s, "Thassa's Oracle") > card_value(s, "Island")
    # Once the combo is assembled, the finish becomes the top tutor target.
    s2 = GameState(battlefield=[
        krark_body("Krark, the Thumbless"),
        krark_body("Sakashima of a Thousand Faces", copy_of="Krark, the Thumbless"),
        Permanent("Storm-Kiln Artist", summoning_sick=False)])
    assert card_value(s2, "Thassa's Oracle", for_tutor=True) >= 70.0
    print("[ok] tutor grabs acceleration early, the finish once combo-ready")

File: rust/src_backup_2026-06-22_overnight/wishlist.rs
//! wishlist.rs — port of wishlist.py. What the deck most wants to draw/fetch given the board.
//! Used by selection (Ponder/Brainstorm), tutors (Gamble/ETB), the mulligan and discard.

use crate::cards::Registry;
use crate::game_state::GameState;

const PAYOFFS: &[&str] = &["Thassa's Oracle", "Grapeshot", "Brain Freeze"];
const BODIES: &[&str] = &[
    "Krark, the Thumbless", "Sakashima of a Thousand Faces", "Glasspool Mimic", "Phantasmal Image",
];
const DOUBLERS: &[&str] = &["Veyran, Voice of Duality", "Harmonic Prodigy"];
const DRAW_ENGINES: &[&str] = &["Archmage Emeritus"];
const MANA_ENGINES: &[&str] = &["Storm-Kiln Artist", "Birgi, God of Storytelling"];
const FAST_MANA: &[&str] = &[
    "Sol Ring", "Mox Diamond", "Chrome Mox", "Lotus Petal", "Arcane Signet", "Springleaf Drum",
    "Lion's Eye Diamond", "Mana Vault", "Mox Amber", "Relic of Legends",
];
const RITUALS: &[&str] = &["Pyretic Ritual", "Desperate Ritual", "Strike It Rich", "Rite of Flame"];

const COMBO_DUALCASTER: &str = "Dualcaster Mage";
const COMBO_SHIMMERS: &[&str] = &["Twinflame", "Heat Shimmer"];

const ENGINE_EXTRA: &[&str] = &[
    "Krark's Thumb", "Baral, Chief of Compliance", "Underworld Breach",
    "Lion's Eye Diamond", "Jeska's Will", "Brainstorm", "Ponder", "Frantic Search",
    "Gamble", "Pyretic Ritual", "Desperate Ritual", "Strike It Rich",
    "Gitaxian Probe", "Peek", "Borne Upon a Wind", "Rite of Flame",
    "Opt", "Consider", "Serum Visions", "Preordain",
];

pub fn is_body(name: &str) -> bool {
    BODIES.contains(&name)
}

const SINK_PERMS_EXTRA: &[&str] = &[
    "Krark's Thumb", "Baral, Chief of Compliance", "Urabrask", "Tavern Scoundrel",
    "Vivi Ornitier", "Valley Floodcaller", "Okaun, Eye of Chaos", "Zndrsplt, Eye of Wisdom",
];

/// Mirror of loops._SINK_PERMS = _BODIES | _DOUBLERS | _DRAW_ENGINES | _MANA_ENGINES | extra.
pub fn is_sink_perm(name: &str) -> bool {
    BODIES.contains(&name)
        || DOUBLERS.contains(&name)
        || DRAW_ENGINES.contains(&name)
        || MANA_ENGINES.contains(&name)
        || SINK_PERMS_EXTRA.contains(&name)
}

pub fn is_payoff(name: &str) -> bool {
    PAYOFFS.contains(&name)
}

/// Mirror of sim._ACTION's wishlist._ENGINE component (in_engine).
pub fn in_engine_pub(name: &str) -> bool {
    in_engine(name)
}

fn in_engine(name: &str) -> bool {
    BODIES.contains(&name)
        || DOUBLERS.contains(&name)
        || DRAW_ENGINES.contains(&name)
        || MANA_ENGINES.contains(&name)
        || PAYOFFS.contains(&name)
        || ENGINE_EXTRA.contains(&name)
}

fn have(s: &GameState, name: &str) -> bool {
    s.hand.iter().any(|c| c == name) || s.has_permanent(name)
}

pub fn payoff_accessible(s: &GameState) -> bool {
    PAYOFFS.iter().any(|p| {
        s.hand.iter().any(|c| c == p) || s.has_permanent(p) || s.graveyard.iter().any(|c| c == p)
    })
}

fn brain_freeze_ready(s: &GameState, reg: &Registry) -> bool {
    s.krark_bodies(reg) >= 2
        && (s.has_permanent("Archmage Emeritus")
            || s.has_permanent("Storm-Kiln Artist")
            || s.has_permanent("Birgi, God of Storytelling"))
}

fn combo_ready(s: &GameState, reg: &Registry) -> bool {
    s.krark_bodies(reg) >= 2
        && (MANA_ENGINES.iter().any(|m| s.has_permanent(m))
            || DRAW_ENGINES.iter().any(|d| s.has_permanent(d))
            || s.has_permanent("Urabrask")
            || s.has_permanent("Tavern Scoundrel")
            || s.has_permanent("Underworld Breach"))
}

fn ready_to_finish(s: &GameState, reg: &Registry, name: &str) -> bool {
    if name == "Brain Freeze" {
        brain_freeze_ready(s, reg)
    } else {
        combo_ready(s, reg)
    }
}

/// Higher = more wanted now. Mirror of wishlist.card_value.
pub fn card_value(s: &GameState, reg: &Registry, name: &str, for_tutor: bool) -> f64 {
    let mut score = 0.0;
    let is_b = is_body(name);
    let accessible = payoff_accessible(s);

    let have_dc = have(s, COMBO_DUALCASTER);
    let have_shimmer = COMBO_SHIMMERS.iter().any(|sh| have(s, sh));
    if name == COMBO_DUALCASTER && have_shimmer && !have_dc {
        score += 120.0;
    }
    if COMBO_SHIMMERS.contains(&name) && have_dc && !have_shimmer {
        score += 120.0;
    }

    let bodies = s.krark_bodies(reg);
    if is_b && (1..2).contains(&bodies) {
        score += 80.0;
    }
    if DOUBLERS.contains(&name) && s.trigger_doublers(reg) == 0 {
        score += 60.0;
    }
    if name == "Krark's Thumb" && !s.has_permanent("Krark's Thumb") {
        score += 50.0;
    }
    if DRAW_ENGINES.contains(&name) && !DRAW_ENGINES.iter().any(|d| s.has_permanent(d)) {
        score += 50.0;
    }
    if MANA_ENGINES.contains(&name) && !MANA_ENGINES.iter().any(|m| s.has_permanent(m)) {
        score += 40.0;
    }
    if name == "Jeska's Will" {
        score += if bodies >= 1 { 60.0 } else { 35.0 };
    } else if FAST_MANA.contains(&name) {
        score += 35.0;
    } else if RITUALS.contains(&name) {
        score += 20.0;
    }

    if PAYOFFS.contains(&name) {
        if !for_tutor {
            score += if name == "Thassa's Oracle" { 45.0 } else { 15.0 };
        }
        if !accessible && ready_to_finish(s, reg, name) {
            score += 100.0;
        }
    }

    if in_engine(name) && !PAYOFFS.contains(&name) {
        score += 10.0;
        if is_b {
            score += 5.0;
        }
    }
    score
}

/// The k highest-value cards from `pool`, best first. Stable like Python's sorted
/// (ties keep original order). Mirror of wishlist.best.
pub fn best(s: &GameState, reg: &Registry, pool: &[String], k: usize, for_tutor: bool) -> Vec<String> {
    let mut idx: Vec<usize> = (0..pool.len()).collect();
    // sort by -value, stable
    idx.sort_by(|&a, &b| {
        let va = card_value(s, reg, &pool[a], for_tutor);
        let vb = card_value(s, reg, &pool[b], for_tutor);
        vb.partial_cmp(&va).unwrap_or(std::cmp::Ordering::Equal)
    });
    idx.into_iter().take(k).map(|i| pool[i].clone()).collect()
}

/// Move the highest-value library card matching `predicate` into hand. Returns the fetched
/// card name, or None. Mirror of wishlist.tutor.
pub fn tutor<F: Fn(&str) -> bool>(s: &mut GameState, reg: &Registry, predicate: F) -> Option<String> {
    let matches: Vec<String> = s.library.iter().filter(|c| predicate(c)).cloned().collect();
    if matches.is_empty() {
        return None;
    }
    let fetched = best(s, reg, &matches, 1, true)[0].clone();
    if let Some(pos) = s.library.iter().position(|c| *c == fetched) {
        s.library.remove(pos);
    }
    s.hand.push(fetched.clone());
    Some(fetched)
}
